File: versioned_docs/version-1.21.x/examples/rs/examples/data_management_attachments.rs
use reduct_rs::{ReductClient, ReductError};
use serde_json::json;
use std::collections::HashMap;
use tokio;

#[tokio::main]
async fn main() -> Result<(), ReductError> {
    // Create a client instance, then get or create a bucket
    let client = ReductClient::builder()
        .url("http://127.0.0.1:8383")
        .api_token("my-token")
        .build();
    let bucket = client
        .create_bucket("my-bucket")
        .exist_ok(true)
        .send()
        .await?;

    // Write attachments for an entry
    bucket
        .write_attachments(
            "camera/front",
            HashMap::from([
                (
                    "schema".to_string(),
                    json!({"type":"object","version":"1.0"}),
                ),
                ("calibration".to_string(), json!({"fx":1260.1,"fy":1261.2})),
            ]),
        )
        .await?;

    // Read attachments
    let attachments = bucket.read_attachments("camera/front").await?;
    assert_eq!(
        attachments.get("schema"),
        Some(&json!({"type":"object","version":"1.0"}))
    );

    // Remove one attachment by key
    bucket
        .remove_attachments("camera/front", Some(vec!["calibration".to_string()]))
        .await?;
    let attachments = bucket.read_attachments("camera/front").await?;
    assert!(!attachments.contains_key("calibration"));

    // Remove all attachments
    bucket.remove_attachments("camera/front", None).await?;
    let attachments = bucket.read_attachments("camera/front").await?;
    assert!(attachments.is_empty());

    Ok(())
}

File: versioned_docs/version-1.21.x/examples/py/src/data_management_attachments.py
import asyncio
from reduct import Client, Bucket


async def main():
    # Create a client instance, then get or create a bucket
    async with Client("http://127.0.0.1:8383", api_token="my-token") as client:
        bucket: Bucket = await client.create_bucket("my-bucket", exist_ok=True)

        # Write attachments for an entry
        await bucket.write_attachments(
            "camera/front",
            {
                "schema": {"type": "object", "version": "1.0"},
                "calibration": {"fx": 1260.1, "fy": 1261.2},
            },
        )

        # Read attachments
        attachments = await bucket.read_attachments("camera/front")
        assert attachments["schema"]["type"] == "object"

        # Remove one attachment by key
        await bucket.remove_attachments("camera/front", ["calibration"])
        attachments = await bucket.read_attachments("camera/front")
        assert "calibration" not in attachments

        # Remove all attachments
        await bucket.remove_attachments("camera/front")
        assert await bucket.read_attachments("camera/front") == {}


if __name__ == "__main__":
    asyncio.run(main())
